lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,306 @@
+// BSD-style license that can be found in the LICENSE file.
+
+/// A formatter and parser for [DateTime] in a fixed format [String] pattern.
+///
+/// For example, calling
+/// `FixedDateTimeCodec('YYYYMMDDhhmmss').decodeToLocal('19960425050322')` has
+/// the same result as calling `DateTime(1996, 4, 25, 5, 3, 22)`.
+///
+/// The allowed characters are
+/// * `Y`	for “calendar year”
+/// * `M`	for “calendar month”
+/// * `D`	for “calendar day”
+/// * `E`	for “decade”
+/// * `C`	for “century”
+/// * `h`	for “clock hour”
+/// * `m`	for “clock minute”
+/// * `s`	for “clock second”
+/// * `S`	for “fractional clock second”
+///
+/// Non-allowed characters in the format [pattern] are ignored when decoding a
+/// string, in this case `YYYY kiwi MM` is the same format string as
+/// `YYYY------MM`. When encoding a datetime, the non-format characters are in
+/// the output verbatim.
+///
+/// Note: this class differs from [DateFormat] in that here, the characters are
+/// treated literally, i.e. the format string `YYY` matching `996` would result in
+/// the same as calling `DateTime(996)`. [DateFormat] on the other hand uses the
+/// specification in https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table,
+/// the format string (or "skeleton") `YYY` specifies only the padding, so
+/// `1996` would be a valid match. This limits it's use to format strings
+/// containing delimiters, as the parser would not know how many digits to take
+/// otherwise.
+class FixedDateTimeFormatter {
+  static const _powersOfTen = [0, 1, 10, 100, 1000, 10000, 100000];
+  static const _validFormatCharacters = [
+    _yearCode,
+    _monthCode,
+    _dayCode,
+    _decadeCode,
+    _centuryCode,
+    _hourCode,
+    _minuteCode,
+    _secondCode,
+    _fractionSecondCode,
+  ];
+  static const _yearCode = 0x59; /*Y*/
+  static const _monthCode = 0x4D; /*M*/
+  static const _dayCode = 0x44; /*D*/
+  static const _decadeCode = 0x45; /*E*/
+  static const _centuryCode = 0x43; /*C*/
+  static const _hourCode = 0x68; /*H*/
+  static const _minuteCode = 0x6D; /*m*/
+  static const _secondCode = 0x73; /*s*/
+  static const _fractionSecondCode = 0x53; /*S*/
+
+  ///Store publicly in case the user wants to retrieve it
+  final String pattern;
+
+  ///Whether to use UTC or the local time zone
+  final bool isUtc;
+  final _blocks = _ParsedFormatBlocks();
+
+  /// Creates a new [FixedDateTimeFormatter] with the provided [pattern].
+  ///
+  /// The [pattern] interprets the characters mentioned in
+  /// [FixedDateTimeFormatter] to represent fields of a `DateTime` value. Other
+  /// characters are not special. If [isUtc] is set to false, the DateTime is
+  /// constructed with respect to the local timezone.
+  ///
+  /// There must at most be one sequence of each special character to ensure a
+  /// single source of truth when constructing the [DateTime], so a pattern of
+  /// `"CCCC-MM-DD, CC"` is invalid, because it has two separate `C` sequences.
```

Comment:
We do allow `"CEEYYY"`, which provides three sources of truth for the year.

Should we try to catch it? (`Y` no longer than 1 if `E` occurs, 2 if `C` occurs, `E` no longer than 1 if `C` occurs.)

Reply:
I think no, as it still is one source of truth. `"CEEYYY"` means exactly what one expects what century, decade, and year mean; `C*100 + E*10 + Y`. Both encoding and decoding can handle this, so I see no unexpected behavior or (hard to handle) ambiguities.

If somebody uses the pattern `"CYYYY"` they should be aware that the resulting year is probably not the same as when using `"YYYY"`, I don't see this as a problem though (especially not as a problem of source of truth).